src/profile-logic/flame-graph.js

```diff
@@ -39,29 +44,135 @@ type Stack = Array<{
+   * conveniently get an array where all its values are uninitialized
+   * from start. */
+
+  let callNodeIndex = 0;
+  let ptr = 0;
+  for (; callNodeIndex < callNodeTable.length; callNodeIndex++) {
+    offsets[callNodeIndex] = ptr;
+    ptr += callNodeChildCount[callNodeIndex];
+
+    if (totalTime[callNodeIndex] === 0) {
+      continue;
+    }
+
+    const parent = callNodeTable.prefix[callNodeIndex];
+    if (parent === -1) {
+      roots.push(callNodeIndex);
+      continue;
+    }
+
+    const funcName = thread.stringTable.getString(
+      thread.funcTable.name[callNodeTable.func[callNodeIndex]]
+    );
+
+    /* From the parent, we can now know the slice allotted for all
+     * its children. */
+    const start = offsets[parent];
+    const end = offsets[parent] + callNodeChildCount[parent] - 1;
+
+    /* Find the place in `array` where this callnode should be
+     * inserted, swapping elements in the array as we go
+     * along. Continue as long as this callnode's function name is
+     * lexically greater than the function names of the callnodes
+     * already placed in the array. This ensures that all slices have
+     * children in descending order. Any callnode indices equal to 0
+     * means that they are uninitialized, so just breeze through
+     * them. When we stop, when have found the right position to
+     * insert our callnode.
+     *
+     * This effectively is an insertion sort, which is O(n^2), but
+     * since n is typically small (the number of children of a given
+     * callnode), it should be just fine.
+     */
+    let i = start;
+    while (
+      i < end &&
+      (array[i + 1] === 0 ||
+        funcName <
+          thread.stringTable.getString(
+            thread.funcTable.name[callNodeTable.func[array[i + 1]]]
+          ))
+    ) {
+      array[i] = array[i + 1];
+      i++;
+    }
+    array[i] = callNodeIndex;
+  }
+  offsets[callNodeIndex] = ptr;
+  return { roots, children: { array, offsets } };
+}
+
 /**
  * Build a FlameGraphTiming table from a call tree.
  */
 export function getFlameGraphTiming(
-  callTree: CallTree.CallTree
+  thread: Thread,
+  interval: Milliseconds,
+  callNodeInfo: CallNodeInfo,
+  invertCallstack: boolean
 ): FlameGraphTiming {
-  callTree.preloadChildrenCache();
-
+  const {
+    callNodeChildCount,
+    callNodeTimes,
+    rootTotalTime,
+  } = computeCallTreeCountsAndTimings(
+    thread,
+    callNodeInfo,
+    interval,
+    invertCallstack
+  );
+  const { roots, children } = getRootsAndChildren(
+    thread,
+    callNodeInfo.callNodeTable,
+    callNodeChildCount,
+    callNodeTimes.totalTime
+  );
   const timing = [];
+
   // Array of call nodes to recursively process in the loop below.
   // Start with the roots of the call tree.
-  const stack: Stack = callTree
-    .getRoots()
-    .map(nodeIndex => ({ nodeIndex, depth: 0 }));
+  const stack: Stack = roots.map(nodeIndex => ({ nodeIndex, depth: 0 }));
```

Comment:
I wouldn't give this type a name, since it's only used internally in this function:

```
  const stack: Array<{
    depth: number,
    nodeIndex: IndexIntoCallNodeTable,
  }> = roots.map(nodeIndex => ({ nodeIndex, depth: 0 }));
```